examples/go-control-plane/adapter.go

```diff
@@ -10,83 +10,251 @@ import (
+				TransportApiVersion:       core.ApiVersion_V3,
+				ApiType:                   core.ApiConfigSource_GRPC,
+				SetNodeOnFirstMessageOnly: true,
+				GrpcServices: []*core.GrpcService{{
+					TargetSpecifier: &core.GrpcService_EnvoyGrpc_{
+						EnvoyGrpc: &core.GrpcService_EnvoyGrpc{ClusterName: clusterName},
+					},
+				}},
+			},
 		}
+	return source
+}
+
+func buildHttpConnectionManager() *hcm.HttpConnectionManager {
+	// HTTP filter configuration.
+	manager := &hcm.HttpConnectionManager{
+		CodecType:  hcm.HttpConnectionManager_AUTO,
+		StatPrefix: "http",
+		HttpFilters: []*hcm.HttpFilter{{
+			Name: wellknown.Router,
+		}},
 	}
+	return manager
+}
 
-	// Parse Listeners
-	listeners := make(map[string]*listener.Listener)
-	if state.Listeners != nil {
+func makeListener(listenerName string, address string, port uint32, filterChains []*listener.FilterChain) *listener.Listener {
+	return &listener.Listener{
+		Name: listenerName,
+		Address: &core.Address{
+			Address: &core.Address_SocketAddress{
+				SocketAddress: &core.SocketAddress{
+					Protocol: core.SocketAddress_TCP,
+					Address:  address,
+					PortSpecifier: &core.SocketAddress_PortValue{
+						PortValue: port,
+					},
+				},
+			},
+		},
+		FilterChains: filterChains,
+	}
+}
 
-		for _, l := range state.Listeners.Items {
-			listeners[l.Name] = &listener.Listener{
-				Name:                             l.Name,
-				Address:                          &core.Address{
-					Address: &core.Address_SocketAddress{
-						SocketAddress: &core.SocketAddress{
-							Address: l.Address,
-						},
+func MakeRouteHTTPListener(clusterName string, listenerName string, listenerAddress string, port uint32, route string) *listener.Listener {
+	rdsSource := configSource(clusterName)
+	routeSpecifier := &hcm.HttpConnectionManager_Rds{
+		Rds: &hcm.Rds{
+			ConfigSource:    rdsSource,
+			RouteConfigName: route,
+		},
+	}
+
+	manager := buildHttpConnectionManager()
+	manager.RouteSpecifier = routeSpecifier
+
+	pbst, err := ptypes.MarshalAny(manager)
+	if err != nil {
+		panic(err)
+	}
+
+	filterChains := []*listener.FilterChain{
+		{
+			Filters: []*listener.Filter{
+				{
+					Name: wellknown.HTTPConnectionManager,
+					ConfigType: &listener.Filter_TypedConfig{
+						TypedConfig: pbst,
 					},
 				},
-			}
-		}
+			},
+		},
+	}
+
+	return makeListener(listenerName, listenerAddress, port, filterChains)
+}
+
+// MakeRuntime creates an RTDS layer with some fields.
+func MakeRuntime(runtimeName string) *runtime.Runtime {
+	return &runtime.Runtime{
+		Name: runtimeName,
+		Layer: &pstruct.Struct{
+			Fields: map[string]*pstruct.Value{
+				"field-0": {
+					Kind: &pstruct.Value_NumberValue{NumberValue: 100},
+				},
+				"field-1": {
+					Kind: &pstruct.Value_StringValue{StringValue: "foobar"},
+				},
+			},
+		},
+	}
+}
+
+func (a *adapterServer) SetState (ctx context.Context, state *pb.Snapshot) (response *pb.SetStateResponse, err error) {
+
+	numClusters := len(state.Clusters.Items)
```

Comment:
I'm unsure if this matters nor the benchmarking on it, I know Go's compiler is quite smart so it might not make a difference but a possibly faster way to write it could be

```go
	var clusters []types.Resource
        for i, cluster := range state.Clusters.Items {
		clusters = append(clusters, MakeCluster(cluster.Name, state.Node))
        }
```

Reply:
it works and is so much shorter! thank you.

Reply:
Oh! Switch to `for _, cluster := range state.Clusters.Items`, I started to use it but it's not needed